[12.0][MIG] account_invoice_line_default_account

SirTakobi opened the pull request:
Migrating/Adding **account_invoice_line_default_account** to `12.0`.

This poor module has a troubled history: as far as I have found, it was brought from launchpad's `6.1` (unmerged) to `8.0` with https://github.com/OCA/account-invoicing/pull/23, then https://github.com/OCA/account-invoicing/pull/189, and finally https://github.com/OCA/account-invoicing/pull/356 (currently still open).

Then a migration to `10.0` happened in https://github.com/OCA/account-invoicing/pull/351 (currently still open).

I tried to put together something like a consistent commit history and this is the result, with a last commit for migrating to `12.0`.

Additional notes:

- I added some tests :tada:
- ~~There is a fixup commit because I couldn't find where the `__openerp__` was removed between https://github.com/OCA/account-invoicing/pull/356 and https://github.com/OCA/account-invoicing/pull/351, I didn't squash it in order not to mix authorships, but I can do it if @jbaudoux agrees~~
- About the roadmap:
  - For `account.invoice.line._onchange_account_id` I already tried substituting `write` with `update` but that results in not updating the partner (can be verified with the tests),
  - For `account.invoice.line.account_id`, currently the module assumes the previous `default` method is the one defined in **account** but that might not be True; retrieving the field's `default` inside the defined method is useless because it has already been changed to the defined method

  If you have any suggestion it is welcome

Looking at the linked PRs and commits, @jbaudoux you might want to have a look :)

NL66278 committed:
[FIX] Update module taking into account reviews bij Stefan Rijnhart (Therp)
            and Alexandre Fayolle (CampToCamp):
            - No hyphenation in module description
            - Better module name
            - Typo's fixed
            - Streamline access to context dictionary: context.get('key')
(../account-financial-tools-6.1 rev 100.1.2)

NL66278 committed:
[FIX] refactored code to make it more mantainable.
(../account-financial-tools-6.1 rev 100.1.3)

Commit:
[ADD] Module to track default expense account on invoice lines per supplier
(../account-financial-tools-6.1 rev 101)

NL66278 committed:
[FIX] Correction of author name.
(../account-financial-tools-6.1 rev 100.1.4)

Commit:
[FIX] account_invoice_line_default_account: Correction of author name.
(../account-financial-tools-6.1 rev 103)

jbaudoux committed:
[MIG] ported module account_invoice_line_default_account to 7.0

jbaudoux committed:
[MIG] account_invoice_line_default_account: Migration to 8.0
[ADD] account_invoice_line_default_account

[IMP] account_invoice_line_default_account: updated with new OCA guidelines

[FIX] pep8, README.rst

[FIX] view_load is deprecated

[FIX] fields.property definition have changed

[FIX] from . import + contributors

[FIX] account_invoice_line_default_account: Adding the partner in the context of invoice line of customer invoice. It's there in standard in supplier invoices but not in customer invoices and it's required by this module

[ADD]account_invoice_line_default_account: default income account on partner similar to default expense account

[ADD] translations fr/nl

[FIX] __openerp__ file

[IMP] look and feel

[FIX] pep8

jbaudoux committed:
[MIG] account_invoice_line_default_account: Migration to 10.0
account_invoice_line_default_account: search based on xmlid as name is translatable
[FIX] pylint

SirTakobi committed:
[MIG] account_invoice_line_default_account: Migration to 12.0
This module is not compatible with account_invoice_line_sequence because them both add to the context of invoice lines in invoice form view

jbaudoux commented on account_invoice_line_default_account/models/account_invoice_line.py:
```diff
@@ -0,0 +1,58 @@
+    account_id = fields.Many2one(default=_account_id_default)
+
+    @api.onchange('account_id')
+    def _onchange_account_id(self):
```
maybe better to use a different name to not have to call super

SirTakobi replied:
Thanks for the review!
I can see no harm in calling `super`, actually it looks to me like a common thing to call `super` when overriding a method; why should this case be treated in a different way?

jbaudoux replied:
Before `api.onchange`, there could be only a single method called on an onchange. With `api.onchange` you can have multiple different methods. The conditions are not to test if super has to be called or not but if this method has to be executed. So I find the current implementation hard to read.
Nevertheless it will work. Just a minor remark.

SirTakobi replied:
That makes perfect sense: overriding the method is really needed only when you want to edit the method's behavior, but that does not have to be the way to go with the `onchange` methods.
I have done this change too, with a bit of explanation in the comments, let me know if it looks better now :)

jbaudoux replied:
Looks clear now

jbaudoux commented on account_invoice_line_default_account/models/account_invoice_line.py:
```diff
@@ -0,0 +1,58 @@
+    def _onchange_account_id(self):
+        if not self.account_id or not self.partner_id or self.product_id:
+            return super(AccountInvoiceLine, self)._onchange_account_id()
+        if self._context.get('journal_id'):
```
self.env.context

SirTakobi replied:
Thanks, done :)

jbaudoux reviewed (commented):
Thanks for adding tests.

Some minor remarks.

I remember when I ported to v10 that lot of code had to be rewritten, that's why I squashed all the old commits. I don't think it is useful to recover all those old commits and I would squash them all into a single commit or at least reduce them a lot.

rafaelbn commented:
/ocabot migration account_invoice_line_default_account

SirTakobi commented:
> I remember when I ported to v10 that lot of code had to be rewritten, that's why I squashed all the old commits. I don't think it is useful to recover all those old commits and I would squash them all into a single commit or at least reduce them a lot.

@jbaudoux I squashed most of the commits, and added the **\_\_openerp__.py** removal in `10.0` migration, let me know if it looks better now.
Just as reference, I have kept the previous history in https://github.com/SirTakobi/account-invoicing/tree/12.0-mig-account_invoice_line_default_account-bkp

Can you update your review? Github is not allowing me to re-request you to do it

> /ocabot migration account_invoice_line_default_account

@rafaelbn thanks, I'm sorry I forgot to do those things

jbaudoux commented on account_invoice_line_sequence/__manifest__.py:
```diff
@@ -16,6 +16,9 @@
     'data': ['views/account_invoice_view.xml',
              'views/report_invoice.xml'],
     'depends': ['account'],
+    'excludes': [
```
Why the 2 modules cannot work together?

SirTakobi replied:
I wrote it in the commit message:
> This module is not compatible with account_invoice_line_sequence because them both add to the context of invoice lines in invoice form view

jbaudoux reviewed (approved):
LGTM

eLBati commented on .travis.yml:
```diff
@@ -23,8 +23,10 @@ env:
+  - TESTS="1" ODOO_REPO="OCA/OCB" EXCLUDE="account_invoice_line_default_account"
+  - TESTS="1" ODOO_REPO="OCA/OCB" EXCLUDE="account_invoice_line_sequence"
+  - TESTS="1" ODOO_REPO="odoo/odoo" MAKEPOT="1" EXCLUDE="account_invoice_line_default_account"
+  - TESTS="1" ODOO_REPO="odoo/odoo" MAKEPOT="1" EXCLUDE="account_invoice_line_sequence"
```
Can runboat be informed about this?

https://runboat.odoo-community.org/api/v1/builds/bf7e8fb88-cda7-4c02-a0c2-5717acb6d8bf/init-log

SirTakobi replied:
Runboat currently does not support conflicting addons, see https://github.com/sbidoul/runboat/issues/36

eLBati committed:
REF account_invoice_line_default_account using _onchange_product_id, which is always executed while adding invoice lines

eLBati committed:
ADD italian translation

eLBati committed:
ADD setup

eLBati reviewed (changes requested):
Proposing https://github.com/SirTakobi/account-invoicing/pull/1 to avoid basing on context and thus incompatibilities

SirTakobi commented:
> Proposing [SirTakobi#1](https://github.com/SirTakobi/account-invoicing/pull/1) to avoid basing on context and thus incompatibilities

@eLBati there are some things that do not really convince me in https://github.com/SirTakobi/account-invoicing/pull/1.

Instead of working on the linked PR, can we go on with the merge of this migration alone? Since we already have enough approvals, this module might finally be merged for the first time! :smile:
Then you will be able to propose that PR directly to `12.0`.
Unless you think there is something in this PR that might be blocking for the merge.

SirTakobi commented:
@OCA/accounting-maintainers can this be merged or do I have to change something? Thanks!

eLBati commented:
> Unless you think there is something in this PR that might be blocking for the merge

Uhm yes, this https://github.com/SirTakobi/account-invoicing/pull/1

SirTakobi commented:
> > Unless you think there is something in this PR that might be blocking for the merge
> 
> Uhm yes, this [SirTakobi#1](https://github.com/SirTakobi/account-invoicing/pull/1)

Now that you explained those changes, I merged them :)

Just for future reference, my doubts were about choosing the onchange on `product_id` field specifically, but we clarified that that onchange is called on any line creation from UI; so the onchange can do the job that was previously (when onchanges were not called on line creation) performed from the line's view using the context.

SirTakobi committed:
[FIX] account_invoice_line_default_account: Do not overwrite product's account set on line

SirTakobi commented:
@eLBati can you update your review now? Thanks!

eLBati reviewed (approved):
Thanks

SirTakobi commented:
@OCA/accounting-maintainers can this be merged? There are a few happy reviewers already :smile:

SirTakobi commented:
@OCA/accounting-maintainers can this be merged? Thanks!

SirTakobi commented:
@OCA/accounting-maintainers can this be merged? Thanks!

sergiocorato commented:
/ocabot merge nobump
